cpp/tests/structs/utilities_tests.cpp

```diff
@@ -0,0 +1,220 @@
+/*
+ * Copyright (c) 2021, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <cudf_test/base_fixture.hpp>
+#include <cudf_test/column_utilities.hpp>
+#include <cudf_test/column_wrapper.hpp>
+#include <cudf_test/cudf_gtest.hpp>
+#include <cudf_test/iterator_utilities.hpp>
+#include <cudf_test/table_utilities.hpp>
+#include <cudf_test/type_lists.hpp>
+
+#include <cudf/detail/aggregation/aggregation.hpp>
+#include <structs/utilities.hpp>
+
+namespace cudf::test {
+
+/**
+ * @brief Round-trip input table through flatten/unflatten,
+ *        verify that the table remains equivalent.
+ */
+void flatten_unflatten_compare(table_view const& input_table)
+{
+  using namespace cudf::structs::detail;
+
+  auto [flattened, _, __, ___] =
+    flatten_nested_columns(input_table, {}, {}, column_nullability::FORCE);
+  auto unflattened =
+    unflatten_nested_columns(std::make_unique<cudf::table>(flattened), input_table);
+
+  CUDF_TEST_EXPECT_TABLES_EQUIVALENT(input_table, unflattened->view());
+}
+
+using namespace cudf;
+using iterators::null_at;
+using strings = strings_column_wrapper;
+using structs = structs_column_wrapper;
+
+struct StructUtilitiesTest : BaseFixture {
+};
+
+template <typename T>
+struct TypedStructUtilitiesTest : StructUtilitiesTest {
+};
+
+TYPED_TEST_CASE(TypedStructUtilitiesTest, FixedWidthTypes);
+
+TYPED_TEST(TypedStructUtilitiesTest, ListsAtTopLevelUnsupported)
+{
+  using T     = TypeParam;
+  using lists = lists_column_wrapper<T, int32_t>;
+  using nums  = fixed_width_column_wrapper<T, int32_t>;
+
+  auto lists_col = lists{{0, 1}, {22, 333}, {4444, 55555, 666666}};
```

Comment:
Be careful with large integer numbers. I suspect that this test does not yield correct results, since `666666` will cause overflow for `int8_t`.

Reply:
Same is applied for the tests below.

Reply:
Fair point. I don't think this really affects the test, but it'd be better to reduce the ranges of the input values.

Reply:
Yeah, maybe overflow numbers are wrapped around so your tests didn't fail. But this is UB :)